cli/cmd/repair.go

```diff
@@ -32,80 +49,121 @@ linkerd upgrade.`,
-			}
+	if !forced {
+		if serverVersion != clientVersion {
 
-			clockSkewDuration, err := time.ParseDuration(values.Identity.Issuer.ClockSkewAllowance)
-			if err != nil {
-				return fmt.Errorf("Failed to parse ClockSkewAllowance from linkerd-config: %s", err)
-			}
-			issuanceLifetime, err := time.ParseDuration(values.Identity.Issuer.IssuanceLifetime)
-			if err != nil {
-				return fmt.Errorf("Failed to parse IssuanceLifetime from linkerd-config: %s", err)
-			}
-			idCtx := pb.IdentityContext{
-				TrustAnchorsPem:    values.IdentityTrustAnchorsPEM,
-				Scheme:             values.Identity.Issuer.Scheme,
-				ClockSkewAllowance: ptypes.DurationProto(clockSkewDuration),
-				IssuanceLifetime:   ptypes.DurationProto(issuanceLifetime),
-				TrustDomain:        values.IdentityTrustDomain,
+			// Suggest directly upgrading to 2.9.4 or above for older versions
+			if repairNotApplicableVersionRegex.Match([]byte(serverVersion)) {
+				return fmt.Errorf("repair command is only applicable to 2.9 control-plane versions. Please try upgrading to the latest supported versions of Linkerd i.e 2.9.4 and above")
 			}
 
-			// Populate identity values
-			err = fetchIdentityValues(cmd.Context(), k8sAPI, &idCtx, &values)
-			if err != nil {
-				return fmt.Errorf("Failed to load issuer credentials: %s", err)
+			// Suggest 2.9.4 CLI version for all 2.9 server versions
+			if repairApplicableVersionRegex.Match([]byte(serverVersion)) {
+				return fmt.Errorf("Please run the repair command with a `stable-2.9.4` CLI.\nRun `export LINKERD2_VERSION=\"stable-2.9.4\"; curl -sL https://run.linkerd.io/install | sh; unset LINKERD2_VERSION` to install the server version of the CLI")
 			}
 
-			valuesMap, err := values.ToMap()
-			if err != nil {
-				return fmt.Errorf("Failed to convert Values into a map: %s", err)
-			}
+			// Suggest server version for everything else. This includes all edge versions
+			return fmt.Errorf("Please run the repair command with a CLI that has the same version as the control plane.\nRun `export LINKERD2_VERSION=\"%s\"; curl -sL https://run.linkerd.io/install | sh; unset LINKERD2_VERSION` to install the server version of the CLI", serverVersion)
+		}
+	}
 
-			overrides, err := renderOverrides(valuesMap, true)
-			if err != nil {
-				return fmt.Errorf("Failed to render overrides: %s", err)
-			}
+	// Load the stored config
+	config, err := k8sAPI.CoreV1().ConfigMaps(controlPlaneNamespace).Get(ctx, "linkerd-config", metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("Failed to load linkerd-config: %s", err)
+	}
 
-			fmt.Print(string(overrides))
+	values := linkerd2.Values{}
 
-			return nil
-		},
+	valuesRaw, ok := config.Data["values"]
+	if !ok {
+		return errors.New("values not found in linkerd-config")
 	}
 
-	return cmd
+	err = yaml.Unmarshal([]byte(valuesRaw), &values)
+	if err != nil {
+		return fmt.Errorf("Failed to load values from linkerd-config: %s", err)
+	}
+
+	// Reset version fields
```

Comment:
No need to add a comment when the function name is explicit enough :wink:

Reply:
This was already there 😁